src/main/java/com/esewa/usermanagement/service/UserServiceImpl.java

```diff
@@ -0,0 +1,66 @@
+package com.esewa.usermanagement.service;
+
+import com.esewa.usermanagement.annotation.HashPassword;
+import com.esewa.usermanagement.constants.ExceptionMessages;
+import com.esewa.usermanagement.entity.User;
+import com.esewa.usermanagement.entity.Wallet;
+import com.esewa.usermanagement.exceptions.UserNotFoundException;
+import com.esewa.usermanagement.repository.UserRepository;
+import com.esewa.usermanagement.repository.WalletRepository;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+
+    private final UserRepository userRepository;
+    private final WalletRepository walletRepository;
+
+    @Override
+    @HashPassword
+    @Transactional
+    @Async("saveMultipleUsersThreadPoolTaskExecutor")
+    public  CompletableFuture<User> registerUser(User user) {
+        System.out.println(Thread.currentThread());
+        Optional<User> userExists = userRepository.findByName(user.getName());
+        if (userExists.isPresent()) {
+            log.error(ExceptionMessages.DUPLICATE_USER + new Date());
+            throw new RuntimeException("Duplicate User: "+ user.getName());
+        }
+        User savedUser = userRepository.save(user);
+        walletRepository.save(new Wallet(0, savedUser.getId()));
+        if (user.getName().equals("shivam")) {
```

Comment:
Can you please add a comment here along the lines of "verify rollback when failure"?

Also, "shivam".equals(user.getName());

Reply:
Done